### PR TITLE
fix: fix Mongo plugin query execution failure 

### LINKED ISSUE #15548: [Bug]: Mongo plugin: queries are erroring out with message "state should be: server session pool is open"
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

- Mongo plugin: queries are erroring out with message "state should be: server session pool is open"
- Ref: https://theappsmith.slack.com/archives/C0341RERY4R/p1659076053299979

### Steps To Reproduce

https://theappsmith.slack.com/archives/C0341RERY4R/p1659076053299979

### Public Sample App

_No response_

### Version

self hosted, cloud

# Temporary fix
- It may be temporarily fixed by either re-saving the datasource or creating a new datasource and linking the queries against it.

### PR DESCRIPTION
**Pull request by sumitsum**
## Description
- Currently some cases have been seen where a Mongo plugin query fails with error message: `state should be: server session pool is open`. 
  - As per my investigation, this occurs because the connection pool at this point has already been closed. Hence running any query against the closed connection pool results in a `IllegalStateException`. 
  - At this point it is not known what is causing the connection pool to close. 
- This PR introduces a change to handle the `IllegalConnectionException` as a `StaleConnectionException` which would result in the connection pool being re-created. 
  - Note: It was not possible to check for closed connection before executing the query because the `MongoClient` object does not seem to expose any `isClosed` kind of method.  

Fixes #15548

## Type of change
- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
- Manual
- JUnit TC

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/server/appsmith-plugins/mongoPlugin/src/main/java/com/external/plugins/MongoPlugin.java [hunk 1]
```diff
@@ -810,6 +819,21 @@ public Mono<DatasourceStructure> getStructure(MongoClient mongoClient, Datasourc
...
+                     * This is to catch the cases when Mongo connection pool closes for some reason and hence throws
+                     * IllegalStateException when query is run.
+                     * Ref: https://github.com/appsmithorg/appsmith/issues/15548
+                     */
+                    .onErrorMap(
+                            IllegalStateException.class,
+                            error -> new StaleConnectionException()
+                    )
+                    // This is an experimental fix to handle the scenario where after a period of inactivity, the mongo
+                    // database drops the connection which makes the client throw the following exception.
```
Comment (nidhi-nair) [inline app/server/appsmith-plugins/mongoPlugin/src/main/java/com/external/plugins/MongoPlugin.java:832]: QQ, where/when are we seeing this error?

Comment (sumitsum) [reply app/server/appsmith-plugins/mongoPlugin/src/main/java/com/external/plugins/MongoPlugin.java:832]: @nidhi-nair this was added earlier (not sure when, would have to check the commit history) to `execute` method. I have justed ported it to `getStructure` as well.

**Review by nidhi-nair — APPROVED** [review]

**Comment by sumitsum** [comment]
@prapullac skipping cypress tests for this : https://theappsmith.slack.com/archives/C02KTH6EHUG/p1659367917217739

**Pull request merged**